akka-remote/src/main/scala/akka/remote/artery/ArteryTcpTransport.scala

```diff
@@ -0,0 +1,286 @@
+    implicit val sys = system
+
+    println(s"# outgoingConnection ${outboundContext.remoteAddress.hostPort}") // FIXME
+    def tcp = Tcp().outgoingConnection(outboundContext.remoteAddress.host.get, outboundContext.remoteAddress.port.get)
+
+    def connectionFlow: Flow[ByteString, ByteString, Future[Tcp.OutgoingConnection]] =
+      if (tlsEnabled) {
+        Flow[ByteString]
+          .map(bytes ⇒ SendBytes(bytes))
+          .viaMat(tls(role = Client).joinMat(tcp)(Keep.right))(Keep.right)
+          .collect {
+            case SessionBytes(_, bytes) ⇒ bytes
+          }
+      } else
+        tcp
+
+    def connectionFlowWithRestart: Flow[ByteString, ByteString, NotUsed] = {
+      import scala.concurrent.duration._
+      // FIXME config of backoff
+      RestartFlow.withBackoff[ByteString, ByteString](1.second, 5.seconds, 0.1) { () ⇒
+        println(s"# RestartFlow ${outboundContext.remoteAddress.hostPort}") // FIXME
+        connectionFlow.mapMaterializedValue(_ ⇒ NotUsed)
+          .log(name = s"outbound connection to [${outboundContext.remoteAddress}]")
+          .addAttributes(Attributes.logLevels(onElement = LogLevels.Off))
+      }
+    }
+
+    Flow[EnvelopeBuffer]
+      .map { env ⇒
+        val bytes = ByteString(env.byteBuffer)
+        bufferPool.release(env)
+        bytes
+      }
+      .recoverWithRetries(1, { case ArteryTransport.ShutdownSignal ⇒ Source.empty })
+      .via(connectionFlowWithRestart)
+      .map(_ ⇒ throw new IllegalStateException(s"Unexpected incoming bytes in outbound connection to [${outboundContext.remoteAddress}]"))
+      .toMat(Sink.ignore)(Keep.right)
+
+    // FIXME The mat value Future is currently never completed, because RestartFlow will retry forever, should it give up?
+    //       related to give-up-system-message-after ?
+
+  }
+
+  override protected def runInboundStreams(): Unit = {
+    implicit val mat = materializer
+    implicit val sys = system
+
+    val controlStream = runInboundControlStream()
+    val ordinaryMessagesStream = runInboundOrdinaryMessagesStream()
+    val largeMessagesStream =
+      if (largeMessageChannelEnabled)
+        runInboundLargeMessagesStream()
+      else
+        Flow[EnvelopeBuffer]
+          .map(_ ⇒ log.warning("Dropping large message, missing large-message-destinations configuration."))
+          .to(Sink.ignore)
+
+    val inboundStream: Sink[EnvelopeBuffer, NotUsed] =
+      Sink.fromGraph(GraphDSL.create() { implicit b ⇒
+        import GraphDSL.Implicits._
+        val partition = b.add(Partition[EnvelopeBuffer](3, env ⇒ {
+          env.byteBuffer.get(EnvelopeBuffer.StreamIdOffset).toInt match {
+            case `ordinaryStreamId` ⇒ 1
+            case `controlStreamId`  ⇒ 0
+            case `largeStreamId`    ⇒ 2
+          }
+        }))
+        partition.out(0) ~> controlStream
+        partition.out(1) ~> ordinaryMessagesStream
+        partition.out(2) ~> largeMessagesStream
```

Comment:
An idea from @jrudolph was that we don't need to include the streamId in the header of each message and have the partition stage here. Instead the very first byte could be the streamId and that is used to connect the inbound connection to the right MergeHub directly.